factor/token/classes/factor.php

```diff
@@ -0,0 +1,251 @@
+<?php
+// This file is part of Moodle - http://moodle.org/
+//
+// Moodle is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+/**
+ * Token factor class.
+ *
+ * @package     factor_token
+ * @author      Peter Burnett <peterburnett@catalyst-au.net>
+ * @copyright   Catalyst IT
+ * @license     http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+
+namespace factor_token;
+
+use tool_mfa\local\factor\object_factor_base;
+use tool_mfa\local\secret_manager;
+
+class factor extends object_factor_base {
+
+    /**
+     * Token implementation.
+     *
+     * {@inheritDoc}
+     */
+    public function has_input() {
+        return false;
+    }
+
+    /**
+     * Token implementation.
+     * This factor is a singleton, return single instance.
+     *
+     * {@inheritDoc}
+     */
+    public function get_all_user_factors($user) {
+        global $DB;
+        $records = $DB->get_records('tool_mfa', array('userid' => $user->id, 'factor' => $this->name));
+
+        if (!empty($records)) {
+            return $records;
+        }
+
+        // Null records returned, build new record.
+        $record = array(
+            'userid' => $user->id,
+            'factor' => $this->name,
+            'timecreated' => time(),
+            'createdfromip' => $user->lastip,
+            'timemodified' => time(),
+            'revoked' => 0,
+        );
+        $record['id'] = $DB->insert_record('tool_mfa', $record, true);
+        return [(object) $record];
+    }
+
+    /**
+     * Token implementation.
+     * Checks whether the user has selected roles in any context.
+     *
+     * {@inheritDoc}
+     */
+    public function get_state() {
+        global $USER;
+
+        // Check if there was a previous locked status to return.
+        $state = parent::get_state();
+        if ($state === \tool_mfa\plugininfo\factor::STATE_LOCKED) {
+            return \tool_mfa\plugininfo\factor::STATE_LOCKED;
+        }
+
+        // Check cookie Exists.
+        if (!NO_MOODLE_COOKIES) {
+            $userid = $USER->id;
+            $cookie = 'MFA_TOKEN_' . $userid;
```

Comment:
```suggestion
            $cookie = 'MFA_TOKEN_' . $USER->id;
```

Should just use the variable directly since it's only used once in this function